gnd/src/main/java/com/google/android/gnd/repository/DataRepository.java

```diff
@@ -112,23 +156,87 @@ private void onProjectLoaded(Project project) {
+   * remote record changes to the local data store before returning the first set. If the network is
+   * not available, relevant records will be returned directly from the local db. After the first
+   * set is emitted, will continue to monitor the remote db for changes, writing updates to records
+   * to the local db and emitting a new set on each change.
+   */
+  public Flowable<ImmutableList<Record>> getRecordSummariesOnceAndStream(
       String projectId, String featureId, String formId) {
     // TODO: Only fetch first n fields.
     // TODO: Also load from db.
     // TODO(#127): Decouple feature from record so that we don't need to fetch record here.
     return getFeature(projectId, featureId)
         .switchIfEmpty(Single.error(new DocumentNotFoundException()))
-        .flatMap(feature -> localDataStore.getRecords(feature))
-        .map(summaries -> filterSummariesByFormId(summaries, formId));
+        .flatMapPublisher(
+            feature -> {
+              Flowable<RemoteDataEvent<Record>> remoteChanges =
+                  remoteDataStore
+                      .loadRecordSummariesOnceAndStreamChanges(feature)
+                      .subscribeOn(Schedulers.io());
+              Flowable<ImmutableList<Record>> localChanges =
+                  localDataStore
+                      .getRecordsOnceAndStream(feature, formId)
+                      .subscribeOn(Schedulers.io());
+              return maybeSyncFirst(remoteChanges)
+                  .toFlowable()
+                  .map(o -> (ImmutableList<Record>) o)
+                  .concatWith(
+                      // This will rewrite the first update in remoteChanges to the local db
+                      // even if already successful. We allow this in case the first update
+                      // timed out or failed or the network wasn't available.
+                      localChanges.mergeWith(
+                          remoteChanges.flatMapCompletable(this::mergeRemoteRecordChange)));
+            });
   }
 
-  private List<Record> filterSummariesByFormId(List<Record> summaries, String formId) {
-    return stream(summaries)
-        .filter(record -> record.getForm().getId().equals(formId))
-        .collect(Collectors.toList());
+  /**
+   * If network if available, wait for first remote emission and update of local store then
+   * complete, otherwise complete immediately. Also completes with success if record sync fails or
+   * times out.
+   */
+  private Completable maybeSyncFirst(Flowable<RemoteDataEvent<Record>> remoteChanges) {
+    if (networkManager.isNetworkAvailable()) {
+      Log.d(TAG, "Has network; syncing records before showing");
+      return remoteChanges
+          .firstElement()
+          .flatMapCompletable(this::mergeRemoteRecordChange)
+          .timeout(GET_REMOTE_RECORDS_TIMEOUT_SECS, TimeUnit.SECONDS)
+          .doOnError(t -> Log.d(TAG, "Record sync timed out"))
```

Comment:
I assume we'll eventually propagate errors from the repository out to callers? e.g. the UI.

Reply:
Good catch (no pun intended).. added clarifying comment/